### PR TITLE
Micropost

### PR DESCRIPTION
**Pull request by iamPhong**

### CODE REVIEW

**Comment by iamPhong** [comment]
Ready

File: app/controllers/microposts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+class MicropostsController < ApplicationController
+  before_action :logged_in_user, only: [:create, :destroy]
```
[the remark this answers is not in the document]
Comment (iamPhong) [reply app/controllers/microposts_controller.rb]: How does an array have an element?

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -14,4 +14,12 @@ def set_locale
   def default_url_options
     {locale: I18n.locale}
   end
+
+  def logged_in_user
+    unless logged_in?
+      store_location
+      flash[:danger] = t("login_err")
```
[the remark this answers is not in the document]
Comment (iamPhong) [reply app/controllers/application_controller.rb]: I don't understand. 
Can you give me the example? pleases...

File: app/views/static_pages/home.html.erb [hunk 3]
```diff
@@ -1,3 +1,19 @@
...
+    <aside class="col-md-4">
+      <section class="user_info">
+        <%= render "shared/user_info" %>
+      </section>
+      <section class="micropost_form">
+        <%= render "shared/micropost_form" %>
+      </section>
+    </aside>
+    <div class="col-md-8">
+      <h3><%= t(".feed_post") %></h3>
```
[the remark this answers is not in the document]
Comment (iamPhong) [reply app/views/static_pages/home.html.erb]: I don't understand.
Can you give me the example? pleases...

**Comment by iamPhong** [comment]
Ready

File: app/controllers/microposts_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdmahmudurrahman) [reply app/controllers/microposts_controller.rb]:
```ruby
before_action :logged_in_user, only: %i(create destroy)
```

File: app/views/static_pages/home.html.erb (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdmahmudurrahman) [reply app/views/static_pages/home.html.erb]:
```ruby
<h3><%= t ".feed_post" %></h3>
```

**Comment by mdmahmudurrahman** [comment]
commented
Please fix all of comments.

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -8,7 +8,9 @@ def index
     @users = User.order(id: :desc).paginate(page: params[:page])
   end
 
-  def show; end
+  def show
+    @microposts = @user.microposts.paginate(page: params[:page])
```
[the remark this answers is not in the document]
Comment (mdmahmudurrahman) [reply app/controllers/users_controller.rb]:
```ruby
before_action :load_user, only: %i(show edit update destroy)
```

File: db/seeds.rb [hunk 5]
```diff
@@ -8,3 +8,9 @@
   User.create!(name:  name, email: email, password: password, password_confirmation: password,
     activated: true, activated_at: Time.zone.now)
 end
+
+users = User.order(:created_at).take(6)
+50.times do
+  content = Faker::Lorem.sentence(5)
+  users.each { |user| user.microposts.create!(content: content) }
+end
```
[the remark this answers is not in the document]
Comment (iamPhong) [reply db/seeds.rb]: I hope it correct... :(

**Comment by iamPhong** [comment]
Ready

File: app/views/shared/_micropost_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<%= form_for(@micropost) do |f| %>
```
Comment (mdmahmudurrahman) [inline app/views/shared/_micropost_form.html.erb]: Remove optional ()

File: app/views/shared/_user_info.html.erb [hunk 7]
```diff
@@ -0,0 +1,4 @@
+<%= link_to gravatar_for(current_user, size: 50), current_user %>
```
Comment (mdmahmudurrahman) [inline app/views/shared/_user_info.html.erb]: Use settings.yml to declare a constant, (here 50)

File: config/routes.rb [hunk 8]
```diff
@@ -11,6 +11,7 @@
     delete "/logout", to: "sessions#destroy"
     resources :users
   end
-    resources :account_activations, only: :edit
-    resources :password_resets, only: [:new, :create, :edit, :update]
+    resources :account_activations, only: %i(edit)
```
Comment (mdmahmudurrahman) [inline config/routes.rb]: For one action. no need to declare an array

File: db/seeds.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (mdmahmudurrahman) [reply db/seeds.rb]: Please remove optional ()

**Comment by mdmahmudurrahman** [comment]
commented

File: db/seeds.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (iamPhong) [reply db/seeds.rb]: Remove optional () is not working

[the remark this answers is not in the document]
Comment (mdmahmudurrahman) [reply db/seeds.rb]:
```ruby
content = Faker::Lorem.sentence 5
```

**Comment by iamPhong** [comment]
Ready

File: app/assets/javascripts/custom.js [hunk 9]
```diff
@@ -0,0 +1,6 @@
+$("#micropost_picture").bind("change", function () {
```
Comment (mdmahmudurrahman) [inline app/assets/javascripts/custom.js]: Please use single quote in js

File: app/controllers/password_resets_controller.rb [hunk 10]
```diff
@@ -1,5 +1,5 @@
 class PasswordResetsController < ApplicationController
-  before_action :[get_user, check_expiration, valid_user] only: [:edit, :update]
+  before_action :get_user, :check_expiration, :valid_user only: %i(edit update)
```
Comment (mdmahmudurrahman) [inline app/controllers/password_resets_controller.rb]: Does it work properly? I find that there is a missing comma.

**Comment by mdmahmudurrahman** [comment]
commented

File: app/controllers/password_resets_controller.rb (hunk 10 above)
Comment (iamPhong) [reply app/controllers/password_resets_controller.rb]: It is working

Comment (iamPhong) [reply app/controllers/password_resets_controller.rb]: but i can edit is like your idea

**Comment by iamPhong** [comment]
Ready

File: app/assets/javascripts/custom.js [hunk 11]
```diff
@@ -0,0 +1,6 @@
+$('#micropost_picture').bind('change', function () {
+  var size_in_megabytes = this.files[0].size / 1024 / 1024;
+  if (size_in_megabytes > 5) {
+    alert(<%= t 'upload_pic_err' %>);
```
Comment (huongnguyenmta) [inline app/assets/javascripts/custom.js]: using `i18n js`

File: app/controllers/microposts_controller.rb [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+      flash[:success] = t "post_created"
+      redirect_to root_path
+    else
+      @feed_items = []
+      render "static_pages/home"
+    end
+  end
+
+  def destroy
+    @micropost.destroy
```
Comment (huongnguyenmta) [inline app/controllers/microposts_controller.rb]: check destroy fail

File: app/models/micropost.rb [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+  mount_uploader :picture, PictureUploader
+  validates :user_id, presence: true
+  validates :content, presence: true, length: {maximum: Settings.post_max_lengh}
+  validate  :picture_size
+
+  private
+
+  def picture_size
+    if picture.size > Settings.picture_size_max.megabytes
+      errors.add :picture, "should be less than 5MB"
```
Comment (huongnguyenmta) [inline app/models/micropost.rb]: I18n

File: app/models/user.rb [hunk 14]
```diff
@@ -54,6 +55,10 @@ def password_reset_expired?
     reset_sent_at < Settings.pass_expired.hours.ago
   end
 
+  def feed
+    Micropost.order_by_desc.where("user_id = ?", id)
```
Comment (huongnguyenmta) [inline app/models/user.rb]: `where("user_id = ?", id)` define scope in model `Micropost`

File: app/models/micropost.rb (hunk 13 above)
Comment (iamPhong) [reply app/models/micropost.rb]: I18n not working here

**Comment by huongnguyenmta** [comment]
commented
[CC][To:1546746] Md. Mahmudur Rahman (Mahmud)
I have some comments in this pull request for you :)

File: app/models/micropost.rb (hunk 13 above)
Comment (mdmahmudurrahman) [reply app/models/micropost.rb]: @iamPhong Use I18n.t

File: app/models/user.rb (hunk 14 above)
Comment (iamPhong) [reply app/models/user.rb]: Em dùng scope trong Micropost : scope :compare_id, -> (id_compare){where("user_id = ?", "%#{id_compare}%")}
cơ mà trong file user.rb em không biết lấy ra cái id của user hiện đang đăng nhập để truyền vào scope kiểu gì cả. Chị chỉ em với @@

File: app/controllers/microposts_controller.rb (hunk 12 above)
Comment (iamPhong) [reply app/controllers/microposts_controller.rb]: ở trên em có before_action :logged_in_user, only: %i(create destroy) mà method logged_in_user ở bên file application_controller.rb em có check fail rồi nên ở đây ko cần nữa ạ

File: app/models/micropost.rb (hunk 13 above)
Comment (iamPhong) [reply app/models/micropost.rb]: I18n không hỗ trợ trong models ạ

Comment (iamPhong) [reply app/models/micropost.rb]: thank you

File: app/models/user.rb [hunk 15]
```diff
@@ -54,6 +55,10 @@ def password_reset_expired?
     reset_sent_at < Settings.pass_expired.hours.ago
   end
 
+  def feed
+    Micropost.order_by_desc.compare_id user.id
```
Comment (iamPhong) [inline app/models/user.rb]: Chính là chỗ này em đang ko biết lấy User id đang đăng nhập kiểu gì ạ

**Comment by iamPhong** [comment]
Ready

File: app/controllers/microposts_controller.rb (hunk 12 above)
Comment (iamPhong) [reply app/controllers/microposts_controller.rb]: I have method logged_in_user in file application_controller.rb
In method logged_in_user have check fail so in method destroy don't need check destroy fail

File: app/models/user.rb (hunk 15 above)
Comment (iamPhong) [reply app/models/user.rb]: I don't know get current user id so it error
Pleasess help me.

File: app/models/micropost.rb [hunk 16]
```diff
@@ -0,0 +1,18 @@
+class Micropost < ApplicationRecord
+  belongs_to :user
+  scope :order_by_desc, -> {order(created_at: :desc)}
+  scope :compare_id, -> (id_compare){where("user_id = ?", "%#{id_compare}%")}
```
Comment (iamPhong) [inline app/models/micropost.rb]: In here, I add scope "compare_id" with parameters is "id_compare" but i don't know get current_user id @@

File: app/models/user.rb (hunk 14 above)
Comment (iamPhong) [reply app/models/user.rb]: it fixed

File: app/models/user.rb (hunk 15 above)
Comment (iamPhong) [reply app/models/user.rb]: it fixed :v

File: app/models/micropost.rb (hunk 16 above)
Comment (iamPhong) [reply app/models/micropost.rb]: it fixed

File: public/javascripts/i18n.js [hunk 17]
```diff
@@ -0,0 +1,1079 @@
+// I18n.js
```
Comment (mdmahmudurrahman) [inline public/javascripts/i18n.js]: Ignore all files from the public folder. You can add it to .gitignore.

File: app/controllers/microposts_controller.rb [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+
+  def micropost_params
+    params.require(:micropost).permit :content, :picture
+  end
+
+  def correct_user
+    @micropost = current_user.microposts.find_by id: params[:id]
+    redirect_to root_url unless @micropost
+  end
+
```
Comment (mdmahmudurrahman) [inline app/controllers/microposts_controller.rb]: Remove this extra new line

**Comment by mdmahmudurrahman** [comment]
commented
Please resolve the comments :D

**Comment by iamPhong** [comment]
Ready

File: app/controllers/microposts_controller.rb [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+    end
+  end
+
+  def destroy
+    if @micropost.destroy
+      flash[:success] = t "post_deleted"
+      redirect_to request.referrer || root_path
+    else
+      flash[:danger] = t "post_deleted_err"
+      redirect_to request.referrer || root_path
```
Comment (mdmahmudurrahman) [inline app/controllers/microposts_controller.rb]: Line 19 and 22 are duplicate. Please remove duplicate code.

File: app/models/micropost.rb [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+  validate  :picture_size
+
+  private
+
+  def picture_size
+    if picture.size > Settings.picture_size_max.megabytes
+      errors.add :picture, I18n.t("file_less_than")
+    end
+  end
+
```
Comment (mdmahmudurrahman) [inline app/models/micropost.rb]: Remove empty line

**Comment by mdmahmudurrahman** [comment]
commented

**Comment by iamPhong** [comment]
Ready
Please check my pull when freetime

**Comment by mdmahmudurrahman** [comment]
(merged)
You can merge the pr.

**Comment by mdmahmudurrahman** [comment]
(merged)
You can merge the pr.

**Comment by mdmahmudurrahman** [comment]
(merged)
You can merge the pr.

**Pull request merged**